server/app/assets/javascripts/file_upload.ts

```diff
@@ -53,6 +56,62 @@ export function init() {
       blockForm.submit()
     }
   })
+
+  const uploadedFilesAttribute = blockForm
+    .querySelector(`[${UPLOADED_FILE_ATTR}]`)
+    ?.getAttribute(UPLOADED_FILE_ATTR)
+
+  if (uploadedFilesAttribute) {
+    const uploadedFilesArray = JSON.parse(uploadedFilesAttribute) as string[]
+    blockForm.addEventListener('formdata', (event) => {
+      const formData = event.formData
+      const file = formData.get('file') as File
+
+      const newName = getUniqueName(file.name, uploadedFilesArray)
+      if (file.name != newName) {
+        // Rename uploaded file, if a file with the same name has already been uploaded.
+        formData.delete('file')
+        formData.append('file', file, newName)
+      }
+    })
+  }
+}
+
+// Exporting so we can test this.
+export function getUniqueName(name: string, existingNames: string[]) {
+  while (existingNames.includes(name)) {
+    const fileNameWithDigitSuffixMatch = name.match(
```

Comment:
suggestion (non-blocking): Can you add a comment with an example of what is returned by `.match`? I looked at the docs and it says an array of matches is returned, but that doesn't make sense with how it's being used here.

Reply:
Added a bunch of comments to hopefully make this more clear. Also found a minor bug (an if statement that is always true) while documenting :)